File: backend/routers/screen_activity.py
import logging
import threading
from typing import List, Optional

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel, Field

import database.screen_activity as screen_activity_db
import database.vector_db as vector_db
from utils.other import endpoints as auth

logger = logging.getLogger(__name__)

router = APIRouter()


class ScreenActivityRow(BaseModel):
    id: int = Field(description="Screenshot ID (used as Firestore document ID)")
    timestamp: str = Field(description="Timestamp in RFC3339 or 'YYYY-MM-DD HH:MM:SS' format")
    appName: str = Field(default='', description="Application name")
    windowTitle: str = Field(default='', description="Window title")
    ocrText: str = Field(default='', description="OCR text from screenshot (truncated to 1000 chars)")
    embedding: Optional[List[float]] = Field(default=None, description="Optional vector embedding (3072-dim Gemini)")


class ScreenActivitySyncRequest(BaseModel):
    rows: List[ScreenActivityRow]


class ScreenActivitySyncResponse(BaseModel):
    synced: int = Field(description="Number of rows written to Firestore")
    last_id: int = Field(description="Maximum row ID from the batch")


@router.post('/v1/screen-activity/sync', response_model=ScreenActivitySyncResponse, tags=['screen-activity'])
def sync_screen_activity(
    request: ScreenActivitySyncRequest,
    uid: str = Depends(auth.get_current_user_uid),
):
    if len(request.rows) > 100:
        raise HTTPException(status_code=400, detail="Maximum 100 rows per batch")

    if not request.rows:
        return ScreenActivitySyncResponse(synced=0, last_id=0)

    # Convert Pydantic models to dicts for database layer
    rows_data = [row.model_dump() for row in request.rows]

    # Firestore upsert (synchronous — blocks response until written)
    try:
        synced = screen_activity_db.upsert_screen_activity(uid, rows_data)
    except Exception:
        logger.exception('Firestore upsert failed for uid=%s', uid)
        raise HTTPException(status_code=500, detail="Failed to sync screen activity")

    # Pinecone vector upsert (fire-and-forget background thread)
    rows_with_embeddings = [r for r in rows_data if r.get('embedding')]
    if rows_with_embeddings:
        thread = threading.Thread(
            target=_upsert_vectors_background,
            args=(uid, rows_with_embeddings),
            daemon=True,
        )
        thread.start()

    last_id = max(row.id for row in request.rows)
    return ScreenActivitySyncResponse(synced=synced, last_id=last_id)


def _upsert_vectors_background(uid: str, rows: list):
    try:
        vector_db.upsert_screen_activity_vectors(uid, rows)
    except Exception:
        logger.exception('Failed to upsert screen activity vectors for uid=%s', uid)

File: desktop/Backend-Rust/src/routes/screen_activity.rs
// Screen Activity sync route
// Receives screenshot metadata + embeddings from the desktop app,
// writes metadata to Firestore and embeddings to Pinecone ns3.

use axum::{
    extract::State,
    http::StatusCode,
    routing::post,
    Json, Router,
};

use crate::auth::AuthUser;
use crate::models::screen_activity::{ScreenActivitySyncRequest, ScreenActivitySyncResponse};
use crate::AppState;

/// POST /v1/screen-activity/sync
async fn sync_screen_activity(
    State(state): State<AppState>,
    user: AuthUser,
    Json(request): Json<ScreenActivitySyncRequest>,
) -> Result<Json<ScreenActivitySyncResponse>, (StatusCode, String)> {
    if request.rows.len() > 100 {
        return Err((StatusCode::BAD_REQUEST, "Maximum 100 rows per batch".to_string()));
    }

    if request.rows.is_empty() {
        return Ok(Json(ScreenActivitySyncResponse { synced: 0, last_id: 0 }));
    }

    let last_id = request.rows.iter().map(|r| r.id).max().unwrap_or(0);

    tracing::info!(
        "Screen activity sync for user {} — {} rows (last_id={})",
        user.uid,
        request.rows.len(),
        last_id
    );

    // Write metadata to Firestore
    let firestore_result = state
        .firestore
        .upsert_screen_activity(&user.uid, &request.rows)
        .await;

    if let Err(e) = &firestore_result {
        tracing::error!("Screen activity Firestore write failed: {}", e);
        return Err((StatusCode::INTERNAL_SERVER_ERROR, format!("Firestore write failed: {}", e)));
    }

    let written = firestore_result.unwrap();

    // Upsert embeddings to Pinecone ns3 (fire-and-forget in background)
    let rows_with_embeddings: Vec<_> = request
        .rows
        .into_iter()
        .filter(|r| r.embedding.is_some())
        .collect();

    if !rows_with_embeddings.is_empty() {
        let config = state.config.clone();
        let uid = user.uid.clone();
        tokio::spawn(async move {
            if let Err(e) = upsert_pinecone_vectors(&config, &uid, &rows_with_embeddings).await {
                tracing::error!("Screen activity Pinecone upsert failed: {}", e);
            }
        });
    }

    Ok(Json(ScreenActivitySyncResponse {
        synced: written,
        last_id,
    }))
}

/// Upsert screen activity embeddings to Pinecone ns3 via REST API.
async fn upsert_pinecone_vectors(
    config: &crate::config::Config,
    uid: &str,
    rows: &[crate::models::screen_activity::ScreenActivityRow],
) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
    let api_key = config
        .pinecone_api_key
        .as_ref()
        .ok_or("PINECONE_API_KEY not configured")?;
    let host = config
        .pinecone_host
        .as_ref()
        .ok_or("PINECONE_HOST not configured")?;

    let vectors: Vec<serde_json::Value> = rows
        .iter()
        .filter_map(|row| {
            let embedding = row.embedding.as_ref()?;
            // Parse timestamp to unix epoch
            let ts = chrono::DateTime::parse_from_rfc3339(&row.timestamp)
                .or_else(|_| chrono::DateTime::parse_from_str(&row.timestamp, "%Y-%m-%d %H:%M:%S"))
                .map(|dt| dt.timestamp())
                .unwrap_or(0);

            Some(serde_json::json!({
                "id": format!("{}-sa-{}", uid, row.id),
                "values": embedding,
                "metadata": {
                    "uid": uid,
                    "screenshot_id": row.id.to_string(),
                    "timestamp": ts,
                    "appName": row.app_name,
                }
            }))
        })
        .collect();

    if vectors.is_empty() {
        return Ok(());
    }

    let client = reqwest::Client::new();
    let url = format!("{}/vectors/upsert", host);

    // Pinecone limit: 100 vectors per upsert
    for chunk in vectors.chunks(100) {
        let body = serde_json::json!({
            "vectors": chunk,
            "namespace": "ns3"
        });

        let resp = client
            .post(&url)
            .header("Api-Key", api_key)
            .header("Content-Type", "application/json")
            .json(&body)
            .send()
            .await?;

        if !resp.status().is_success() {
            let error_text = resp.text().await?;
            tracing::error!("Pinecone upsert error: {}", error_text);
            return Err(format!("Pinecone upsert failed: {}", error_text).into());
        }
    }

    tracing::info!(
        "Screen activity Pinecone upsert uid={} count={}",
        uid,
        vectors.len()
    );
    Ok(())
}

pub fn screen_activity_routes() -> Router<AppState> {
    Router::new().route("/v1/screen-activity/sync", post(sync_screen_activity))
}
